storm-core/src/jvm/org/apache/storm/container/cgroup/CgroupCenter.java

```diff
@@ -0,0 +1,232 @@
+             BufferedReader br = new BufferedReader(reader)) {
+            String str = null;
+            while ((str = br.readLine()) != null) {
+                String[] strSplit = str.split(" ");
+                if (!strSplit[2].equals("cgroup")) {
+                    continue;
+                }
+                String name = strSplit[0];
+                String type = strSplit[3];
+                String dir = strSplit[1];
+                Hierarchy h = hierarchies.get(type);
+                h = new Hierarchy(name, CgroupUtils.analyse(type), dir);
+                hierarchies.put(type, h);
+            }
+            return new ArrayList<Hierarchy>(hierarchies.values());
+        } catch (Exception e) {
+            LOG.error("Get hierarchies error {}", e);
+        }
+        return null;
+    }
+
+    @Override
+    public Set<SubSystem> getSubSystems() {
+
+        Set<SubSystem> subSystems = new HashSet<SubSystem>();
+
+        try (FileReader reader = new FileReader(Constants.CGROUP_STATUS_FILE);
+             BufferedReader br = new BufferedReader(reader)){
+            String str = null;
+            while ((str = br.readLine()) != null) {
+                String[] split = str.split("\t");
+                SubSystemType type = SubSystemType.getSubSystem(split[0]);
+                if (type == null) {
+                    continue;
+                }
+                subSystems.add(new SubSystem(type, Integer.valueOf(split[1]), Integer.valueOf(split[2])
+                        , Integer.valueOf(split[3]).intValue() == 1 ? true : false));
+            }
+            return subSystems;
+        } catch (Exception e) {
+            LOG.error("Get subSystems error {}", e);
+        }
+        return null;
+    }
+
+    @Override
+    public boolean enabled(SubSystemType subsystem) {
+
+        Set<SubSystem> subSystems = this.getSubSystems();
+        for (SubSystem subSystem : subSystems) {
+            if (subSystem.getType() == subsystem) {
+                return true;
+            }
+        }
+        return false;
+    }
+
+    @Override
+    public Hierarchy busy(SubSystemType subsystem) {
+        List<Hierarchy> hierarchies = this.getHierarchies();
+        for (Hierarchy hierarchy : hierarchies) {
+            for (SubSystemType type : hierarchy.getSubSystems()) {
+                if (type == subsystem) {
+                    return hierarchy;
+                }
+            }
+        }
+        return null;
+    }
+
+    @Override
+    public Hierarchy busy(List<SubSystemType> subSystems) {
+        List<Hierarchy> hierarchies = this.getHierarchies();
+        for (Hierarchy hierarchy : hierarchies) {
+            Hierarchy ret = hierarchy;
+            for (SubSystemType subsystem : subSystems) {
+                if (!hierarchy.getSubSystems().contains(subsystem)) {
+                    ret = null;
+                    break;
+                }
+            }
```

Comment:
So if the last Hierarchy element in heirarchies has a set of subsystems that doesn't contain all of `subSystems`, then this 'center' is busy?

Why loop through all the hierarchies?